### PR TITLE
monitoring: Add monitoring architecture doc

### PR DESCRIPTION
**Pull request by anmolbabu**
Add monitoring architecture doc

Issue: https://github.com/Tendrl/documentation/issues/40
Jira: https://tendrl.atlassian.net/browse/TEN-49, https://github.com/Tendrl/documentation/issues/32, https://tendrl.atlassian.net/browse/TEN-84
Signed-off-by: anmolbabu <anmolbudugutta@gmail.com>

### CODE REVIEW

**Comment by anmolbabu** [comment]
@nthomas-redhat @shtripat @brainfunked @anupnivargi @anupnivargi Please review

**Review by brainfunked — CHANGES_REQUESTED** [review]
In the diagram, I would suggest colour coding the boxes to depict the stacks the boxes belong to. In this scenario core and performance monitoring stacks would need to be depicted with their specific colours. The other boxes could be left colourless.

File: tendrl-monitoring-architecture.adoc [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+* The Store contains:
+** Storage System State Information in the Storage System’s own representation,
+   without a pre-defined schema.
+** Storage System State Definition, which is a YAML file that describes the
+   access methods, data types and the general layout of the State Information.
+** Operations hierarchy, which is a Job Queue. This also includes updates
+   regarding the Operations being executed.
+** Status of every relevant Tendrl and Storage System process/component that
+   needs to be monitored.
+** Current time series data needed for automated conditional triggers.
```
Comment (shtripat) [inline tendrl-monitoring-architecture.adoc]: Is it correct?

File: tendrl-monitoring-architecture.adoc [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+* For a host to be managed by Tendrl, an Agent has to run on that host and it
+  must be able to convey it’s own status to the Central Store.
+* Agents are responsible for all operations local to the host they’re on, such
+  as:
+** Hardware interface via an external tool/library. This includes local disk
+   management.
+** Process monitoring for the various system and Storage System daemons,
+   including the Tendrl Bridge via systemd.
+** Invocation of host-local Provisioning Operations.
+** Monitoring and gathering of the system’s status, including collection of
```
Comment (shtripat) [inline tendrl-monitoring-architecture.adoc]: node agent does gathering of time series data??

Comment (brainfunked) [reply tendrl-monitoring-architecture.adoc]: This was an older document, which needs to be updated with the appropriate details from the newly finalised monitoring architecture.

Comment (shtripat) [reply tendrl-monitoring-architecture.adoc]: @anmolbabu can you please update this?

**Comment by anmolbabu** [comment]
@brainfunked @shtripat @nthomas-redhat @anupnivargi I have incorporated comments. Please review.

File: tendrl-monitoring-architecture.adoc [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+the state information in tendrl’s central store for its functioning. Typically
+it can be co-resident with tendrl application but nothing necessiates this.
+Following are its sub-components based on the functionalities:
+
+* Api Layer : This layer will serve the following responsibilities.
+  ** Serve the time series data stored in time series database.
+  ** Alert subscription management for users in tendrl core(This functionality
+     will be achieved for users in tendrl core. These subscriptions will be
+     stored in tendrl’s central store).
+  ** A BSD Licensed python tool by name flask is used for serving the
```
Comment (brainfunked) [inline tendrl-monitoring-architecture.adoc]: This doesn't need to be part of the design discussion. Perhaps can be explained in a separate section on the selection of tools.

File: tendrl-monitoring-architecture.adoc [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+  ** At the start of tendrl performance monitoring application, default
+     monitoring configurations will be loaded to etcd.
+  ** The node-agent and the sds-bridge on any node started after this,
+     will use these default configurations for configuring collectd's plugins
+     (node-agent for physical resource plugins and bridge for bridge specific).
+    *** If the user chose intially not to have monitoring but later decided to
+        have it or node-agent/bridge was started before starting performance
+        monitoring application, a flow will be intiated internally to generate
+        collectd configurations once the performance monitoring application is
+        started.
```
Comment (brainfunked) [inline tendrl-monitoring-architecture.adoc]: The monitoring application should be treated as an add-on stack. This means that the corresponding code, templates, configurations, flows etc. all need to be shipped for the monitoring stack should be shipped separately, with an assumption that the core stack is already started.

I would suggest that the flow framework from the core stack be reused for deploying the monitoring stack.
1. Install the monitoring application.
2. Monitoring application connects to etcd and uploads the default monitoring configuration and flows.
3. Monitoring application triggers the monitoring provisioning flow on every node for node monitoring via the core API.
4. The provisioning flow for the monitoring stack installs the collectd-related pieces on every node and generates the configuration files from templates shipped with the collectd plugins on nodes.

**Review by brainfunked — CHANGES_REQUESTED** [review]

File: api/monitoring-api.adoc [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+The document describes the procedure for fetching statistics from time series
+database.
+
+To explore the resources that are monitored, we can use the api described
+below.
+
+Sample Request
+
+----------
+GET /moitoring/nodes/<node_name>/monitored_types
```
Comment (fbalak) [inline api/monitoring-api.adoc]: There is probably typing error: monitoring instead of moitoring.

Comment (anmolbabu) [reply api/monitoring-api.adoc]: Yes, that was a typo :) .. Thanks for pointing..

File: tendrl-monitoring-architecture.adoc (hunk 3 above)
Comment (anmolbabu) [reply tendrl-monitoring-architecture.adoc]: Done. Removed it from here..

File: tendrl-monitoring-architecture.adoc (hunk 4 above)
Comment (anmolbabu) [reply tendrl-monitoring-architecture.adoc]: Done

**Comment by anmolbabu** [comment]
@brainfunked @nthomas-redhat @shtripat @r0h4n @fbalak Comments incorporated. Please review

**Pull request closed without merging**